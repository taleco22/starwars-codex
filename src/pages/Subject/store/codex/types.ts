/**
 * Action Types
 */

export enum CodexTypes {
  SET_DATA = "data/set_data",
  FAILURE = "data/failure",
  IS_LOADING = "data/is_loading",
  ASYNC_GET_DATA = "data/async_get_data",
}

/**
 * Data Types
 */

export interface Codex {
  count: number
  next: string | null
  previous: string | null
  results: any[]
}

/**
 * State Types
 */

export interface CodexState {
  data: Codex
  isLoading: boolean
}

export interface CodexAction {
  type: string
  payload: CodexState
}
