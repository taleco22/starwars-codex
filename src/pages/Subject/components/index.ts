import Buttons from "./Buttons/Buttons";
import Characters from "./Characters/Characters";
import Films from "./Films/Films";
import Planets from "./Planets/Planets";
import Species from "./Species/Species";
import Starships from "./Starships/Starships";
import Vehicles from "./Vehicles/Vehicles";

export { Buttons, Characters, Films, Planets, Species, Starships, Vehicles };
