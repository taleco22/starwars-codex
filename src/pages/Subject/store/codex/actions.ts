import { Action } from "redux";
import { action } from "typesafe-actions";
import { Codex, CodexTypes } from "./types";

export const triggerLoading = (): Action => action(CodexTypes.IS_LOADING);

export const getAsyncData = (urlPath: string) => action(CodexTypes.ASYNC_GET_DATA, { urlPath });

export const setData = (data: Codex) => action(CodexTypes.SET_DATA, { data });

export const failedRequest = (error: Error) =>
  action(CodexTypes.FAILURE, undefined, undefined, error);
