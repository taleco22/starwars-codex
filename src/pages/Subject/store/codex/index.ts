import { Reducer } from "redux";
import { CodexAction, CodexState, CodexTypes } from "./types";

export const initialState: CodexState = {
  data: {
    count: 0,
    previous: null,
    next: null,
    results: [{}, {}, {}, {}, {}],
  },
  isLoading: true,
};

export const codexReducer: Reducer<CodexState, CodexAction> = (state = initialState, action) => {
  switch (action.type) {
    case CodexTypes.SET_DATA:
      return {
        ...state,
        data: action.payload.data,
        isLoading: false,
      };
    case CodexTypes.IS_LOADING:
      return initialState;
    default:
      return initialState;
  }
};
