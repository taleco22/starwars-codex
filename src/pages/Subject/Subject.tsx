import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { RouteComponentProps } from "react-router-dom";
import { ApplicationState } from "../../redux";
import { getAsyncData, triggerLoading } from "../../redux/ducks/codex/actions";
import { waitButtons } from "../../utils";
import { SelectComponent, SParams } from "../types";
import { Buttons, Characters, Films, Planets, Species, Starships, Vehicles } from "./components";
import { List, Role, Spinner } from "./styled";
export const selector = (state: ApplicationState) => state.codex;

const Subject: React.FC<RouteComponentProps<SParams>> = ({ match }) => {
  const {
    data: { results, count },
    isLoading,
  } = useSelector(selector);
  const dispatch = useDispatch();

  const theme = match.params.theme;

  useEffect(() => {
    waitButtons();
    dispatch(triggerLoading());
    dispatch(getAsyncData(theme));
  }, [dispatch, theme]);

  const Which = () => {
    const selectedComponent: SelectComponent = {
      people: <Characters results={results} />,
      planets: <Planets results={results} />,
      vehicles: <Vehicles results={results} />,
      starships: <Starships results={results} />,
      species: <Species results={results} />,
      films: <Films results={results} />,
    };

    return selectedComponent[theme];
  };

  return (
    <>
      <Role>{theme}</Role>

      {isLoading && <Spinner />}

      <List>{results && Which()}</List>

      {count && <Buttons theme={theme} />}
    </>
  );
};
export default Subject;
