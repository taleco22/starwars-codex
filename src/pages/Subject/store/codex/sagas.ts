import { call, put, takeLatest } from "redux-saga/effects";
import { swapi } from "../../services/api";
import { failedRequest, setData } from "./actions";
import { CodexTypes } from "./types";

export function codexWatcher() {
  return takeLatest(CodexTypes.ASYNC_GET_DATA, getData);
}

interface Action {
  type: string;
  payload: { urlPath: string };
}

export function* getData(action: Action) {
  try {
    const jsonResponse = yield call(swapi, action.payload.urlPath);
    yield put(setData(jsonResponse));
  } catch (e) {
    yield put(failedRequest());
  }
}
