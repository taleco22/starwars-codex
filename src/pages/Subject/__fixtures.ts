import configureStore from "redux-mock-store";
import { RunSagaOptions, stdChannel } from "redux-saga";
import { ApplicationState } from "./store";
import { Codex, CodexState } from "./store/codex/types";

declare let global: { fetch: jest.Mock<any, any> };

export const fetchedData: Codex = {
  count: 10,
  next: "https://swapi.dev/api/people/",
  previous: "https://swapi.dev/api/people/",
  results: [
    {
      name: "Obi-Wan Kenobi",
      height: "182",
      mass: "77",
      hair_color: "auburn, white",
      skin_color: "fair",
      eye_color: "blue-gray",
      birth_year: "57BBY",
      gender: "male",
      homeworld: "https://swapi.dev/api/planets/20/",
      films: [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/6/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/1/",
      ],
      species: ["https://swapi.dev/api/species/1/"],
      vehicles: ["https://swapi.dev/api/vehicles/38/"],
      starships: [
        "https://swapi.dev/api/starships/48/",
        "https://swapi.dev/api/starships/59/",
        "https://swapi.dev/api/starships/64/",
        "https://swapi.dev/api/starships/65/",
        "https://swapi.dev/api/starships/74/",
      ],
      created: "2014-12-10T16:16:29.192000Z",
      edited: "2014-12-20T21:17:50.325000Z",
      url: "https://swapi.dev/api/people/10/",
    },
  ],
};

export const loadingData: Codex = {
  count: 1,
  previous: null,
  next: null,
  results: [{}],
};

export const mockFetchPromise = Promise.resolve({ json: () => Promise.resolve(fetchedData) });

global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);

export const mockGlobal = global;

export const dispatchedActions: any[] = [];

export const channel = stdChannel();

export const fakeStore: RunSagaOptions<any, any> = {
  channel,
  getState: () => "initial",
  dispatch: (action) => dispatchedActions.push(action),
};

const reducer = (state: CodexState) => ({ codex: state });

export const mockStore = configureStore<ApplicationState>()(
  reducer({ data: fetchedData, isLoading: false })
);
export const mockEmpty = configureStore<ApplicationState>()(
  reducer({ data: loadingData, isLoading: true })
);
