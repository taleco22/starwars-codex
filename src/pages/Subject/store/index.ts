/* eslint-disable no-undef */
import { applyMiddleware, compose, createStore, Store } from "redux"
import createSaga from "redux-saga"
import { CodexState } from "./codex/types"
import combinedReducers from "./rootReducer"
import sagas from "./rootSaga"

export interface ApplicationState {
  codex: CodexState
}

declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__?: typeof compose
  }
}

// const dev = window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__();
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose

const sagaMiddleware = createSaga()

const store: Store<ApplicationState> = createStore(
  combinedReducers,
  composeEnhancers(applyMiddleware(sagaMiddleware)),
)

sagaMiddleware.run(sagas)

export default store
