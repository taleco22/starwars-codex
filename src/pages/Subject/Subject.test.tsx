// @ts-nocheck
import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import * as React from "react";
import * as redux from "react-redux";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { ApplicationState } from "../../redux";
import { CodexState } from "../../redux/ducks/codex/types";
import Subject, { selector } from "./Subject";
const mockDispatch = jest.fn();

jest.mock("react-redux", () => ({
  ...jest.requireActual("react-redux"),
  useDispatch: () => mockDispatch,
}));

const mockGoBack = jest.fn();
const mockPush = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useHistory: () => ({
    goBack: mockGoBack,
    push: mockPush,
  }),
}));

const use_dispatcher = jest.spyOn(redux, "useDispatch");

const initial_state = {
  data: {
    count: 10,
    previous: "previous",
    next: "next",
    results: [{ name: "people", gender: "male", mass: "12", height: "12" }],
  },
  isLoading: false,
};

const nullState = {
  data: { count: 1, previous: null, next: null, results: [{}] },
  isLoading: false,
};

const mockMatch = {
  params: {
    theme: "people",
  },
};
const history = jest.fn();
const location = jest.fn();

const reducer = (state: CodexState) => ({ codex: state });
const mockStore = configureStore<ApplicationState>()(reducer(initial_state));
const mockEmpty = configureStore<ApplicationState>()(reducer(nullState));

let use_effect: jest.spyOn<typeof React, "useEffect">;

describe("Subject.tsx", () => {
  it("take snapshot", () => {
    const snap = renderer
      .create(
        <redux.Provider store={mockStore}>
          <Subject history={history} location={location} match={mockMatch} />
        </redux.Provider>
      )
      .toJSON();
    expect(snap).toMatchSnapshot();
  });

  let result: { container: HTMLElement };

  beforeEach(async () => {
    use_effect = jest.spyOn(React, "useEffect");

    use_effect.mockImplementationOnce((f) => f());

    result = render(
      <redux.Provider store={mockStore}>
        <Subject history={history} location={location} match={mockMatch} />
      </redux.Provider>
    );
  });

  afterEach(() => {
    use_dispatcher.mockClear();
    use_effect.mockClear();
    mockDispatch.mockClear();
  });

  it("render component correctly", () => {
    expect(use_effect).toBeCalled();
    expect(use_dispatcher).toBeCalled();
    expect(mockDispatch).toBeCalledTimes(2);
  });

  it("get data from selector", () => {
    expect(use_dispatcher).toBeCalled();
    expect(selector(mockStore.getState()));
  });
});
