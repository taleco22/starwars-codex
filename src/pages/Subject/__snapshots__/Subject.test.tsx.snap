// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Subject.tsx snapshots render with Spinner 1`] = `
Array [
  <h1
    className="sc-fzoLag aoduX"
  >
    people
  </h1>,
  <div
    className="sc-fzoXzr gjwfCR"
  />,
  <ul
    className="sc-AxirZ sc-fzqBZW fYrObk"
  >
    <li
      className="sc-AxiKw sc-fzplWN hGswWD"
    >
      <button
        className="sc-AxhCb sc-fznyAO hmAjLB"
      >
        <h1 />
        s
      </button>
      <div
        className="sc-fznZeY dDFrgx"
      >
        <div
          className="sc-fznKkj iTpAqQ"
          onClick={[Function]}
        />
        <h1 />
        <p>
          gender: 
        </p>
        <p>
          weight: 
          kg
        </p>
        <p>
          height: 
          cm
        </p>
      </div>
    </li>
  </ul>,
  <footer
    className="sc-AxmLO fZfazS"
  >
    <button
      className="sc-AxhCb sc-Axmtr gyNiSL"
      onClick={[Function]}
    >
      Go Back
    </button>
    <button
      className="sc-AxhCb sc-Axmtr gyNiSL"
      onClick={[Function]}
    >
      Next
    </button>
    <div
      className="sc-fzozJi dshpyh"
    >
      <button
        className="sc-AxhCb sc-Axmtr fIjFPo"
        onClick={[Function]}
        tabIndex={0}
      >
        1
      </button>
    </div>
  </footer>,
]
`;

exports[`Subject.tsx snapshots render with data 1`] = `
Array [
  <h1
    className="sc-fzoLag aoduX"
  >
    people
  </h1>,
  <ul
    className="sc-AxirZ sc-fzqBZW fYrObk"
  >
    <li
      className="sc-AxiKw sc-fzplWN hGswWD"
    >
      <button
        className="sc-AxhCb sc-fznyAO hmAjLB"
      >
        <h1>
          Obi-Wan Kenobi
        </h1>
        s
      </button>
      <div
        className="sc-fznZeY dDFrgx"
      >
        <div
          className="sc-fznKkj iTpAqQ"
          onClick={[Function]}
        />
        <h1>
          Obi-Wan Kenobi
        </h1>
        <p>
          gender: 
          male
        </p>
        <p>
          weight: 
          77
          kg
        </p>
        <p>
          height: 
          182
          cm
        </p>
      </div>
    </li>
  </ul>,
  <footer
    className="sc-AxmLO fZfazS"
  >
    <button
      className="sc-AxhCb sc-Axmtr gFtPZS"
      onClick={[Function]}
    >
      Previous
    </button>
    <button
      className="sc-AxhCb sc-Axmtr gFtPZS"
      onClick={[Function]}
    >
      Next
    </button>
    <div
      className="sc-fzozJi dshpyh"
    >
      <button
        className="sc-AxhCb sc-Axmtr fIjFPo"
        onClick={[Function]}
        tabIndex={0}
      >
        1
      </button>
    </div>
  </footer>,
]
`;
